### PR TITLE
Allow "*" to match scheme of protected resource

### PR DESCRIPTION
**Pull request by shekyan**
as discussed in #104 .

### CODE REVIEW

File: index.src.html [hunk 1]
```diff
@@ -2694,11 +2695,15 @@ <h5 id="match-url-to-source-expression" algorithm>
...
 
-  1.  If |expression| is the string "*", and |url|'s {{URL/scheme}} is a
-      <a>network scheme</a>, return "`Matches`". 
+  1.  If |expression| is the string "*", return "`Matches`" if one or more of
+      the following conditions is met:
+      1. |url|'s {{URL/scheme}} is a <a>network scheme</a>.
+
+      2. |url|'s {{URL/scheme}} is the same as |origin|'s {{URL/scheme}}. 
 
       Note: This logic means that in order to allow resource from non-<a>network scheme</a>,
```
Comment (dbates-wk) [inline index.src.html]: I know that this part of the sentence is not in your patch. It does not read well. Maybe something like, "Note: This logic means that in order to allow a resource from a non-network scheme, ..."

File: index.src.html [hunk 2]
```diff
@@ -2694,11 +2695,15 @@ <h5 id="match-url-to-source-expression" algorithm>
...
-      it has to be explicitly whitelisted: `default-src * data: custom-scheme-1: custom-scheme-2:`.
+      it has to be either explicitly whitelisted: `default-src * data: custom-scheme-1: custom-scheme-2:`,
+      or protected resource should be loaded from that same scheme.
```
Comment (dbates-wk) [inline index.src.html]: This part of the sentence does not read well and could be worded strongly. Maybe "... or the protected resource must be loaded from the same scheme."

File: index.src.html [hunk 3]
```diff
@@ -2694,11 +2695,15 @@ <h5 id="match-url-to-source-expression" algorithm>
   |expression| should be resolved. "`'self'`", for instance, will have distinct
   meaning depending on that bit of context.
 
-  1.  If |expression| is the string "*", and |url|'s {{URL/scheme}} is a
-      <a>network scheme</a>, return "`Matches`". 
+  1.  If |expression| is the string "*", return "`Matches`" if one or more of
+      the following conditions is met:
+      1. |url|'s {{URL/scheme}} is a <a>network scheme</a>.
+
+      2. |url|'s {{URL/scheme}} is the same as |origin|'s {{URL/scheme}}.
```
Comment (dbates-wk) [inline index.src.html]: I'm assuming "origin" is analogous to "protected resource". We should standardize on the terminology.

File: index.src.html [hunk 4]
```diff
@@ -419,8 +419,9 @@ <h3 id="changes-from-level-2">Changes from Level 2</h3>
       in [[#unsafe-hashed-attributes-usage]].
 
   9.  The <a>source expression</a> matching has been changed to require explicit whitelisting
-      of any non-<a>network scheme</a>, rather than <a>local scheme</a>, as described
-      in [[#match-url-to-source-expression]].
+      of any non-<a>network scheme</a>, rather than <a>local scheme</a>,
+      unless that non-<a>network scheme</a> is the same as origin scheme, as described
```
Comment (dbates-wk) [inline index.src.html]: I'm assuming "origin" is analogous to "protected resource". We should standardize on the terminology.

File: index.src.html (hunk 1 above)
Comment (shekyan) [inline index.src.html]: Well, it was part of my other patch:) .

File: index.src.html (hunk 2 above)
Comment (shekyan) [inline index.src.html]: Agree.

File: index.src.html (hunk 3 above)
Comment (shekyan) [inline index.src.html]: `origin` is passed to this algorithm, so it's out of scope of this algorithm to define it. But, I guess the caller (https://w3c.github.io/webappsec-csp/#match-url-to-source-list) should define it ? cc/ @mikewest

File: index.src.html (hunk 4 above)
Comment (shekyan) [inline index.src.html]: Let's explicitly say here that it is the origin of protected resource.

**Comment by shekyan** [comment]
Addressed comments.

File: index.src.html [hunk 5]
```diff
@@ -2694,11 +2695,15 @@ <h5 id="match-url-to-source-expression" algorithm>
...
+      1. |url|'s {{URL/scheme}} is a <a>network scheme</a>.
 
-      Note: This logic means that in order to allow resource from non-<a>network scheme</a>,
-      it has to be explicitly whitelisted: `default-src * data: custom-scheme-1: custom-scheme-2:`.
+      2. |url|'s {{URL/scheme}} is the same as |origin|'s {{URL/scheme}}. 
+
+      Note: This logic means that in order to allow resource from a non-<a>network scheme</a>,
+      it has to be either explicitly whitelisted: `default-src * data: custom-scheme-1: custom-scheme-2:`,
+      or the protected resource must be loaded from the same scheme.
       In other words, there is no semantic representation of most permissive |expression|.
```
Comment (mikewest) [inline index.src.html]: I'd just drop this line. I don't think it adds anything.

File: index.src.html [hunk 6]
```diff
@@ -2694,11 +2695,15 @@ <h5 id="match-url-to-source-expression" algorithm>
   |expression| should be resolved. "`'self'`", for instance, will have distinct
   meaning depending on that bit of context.
 
-  1.  If |expression| is the string "*", and |url|'s {{URL/scheme}} is a
-      <a>network scheme</a>, return "`Matches`". 
+  1.  If |expression| is the string "*", return "`Matches`" if one or more of
+      the following conditions is met:
+      1. |url|'s {{URL/scheme}} is a <a>network scheme</a>.
```
Comment (mikewest) [inline index.src.html]: Nit: Newline before this.

File: index.src.html [hunk 7]
```diff
@@ -2694,11 +2695,15 @@ <h5 id="match-url-to-source-expression" algorithm>
...
 
-      Note: This logic means that in order to allow resource from non-<a>network scheme</a>,
-      it has to be explicitly whitelisted: `default-src * data: custom-scheme-1: custom-scheme-2:`.
+      2. |url|'s {{URL/scheme}} is the same as |origin|'s {{URL/scheme}}.
```
Comment (mikewest) [inline index.src.html]: Nit: Newline before this.

File: index.src.html (hunk 3 above)
Comment (mikewest) [inline index.src.html]: Yeah. This is an oversight... I guess I refactored it at some point and missed the caller. I'll fix that.

Comment (mikewest) [inline index.src.html]: I took a stab at reworking this in 3739f6f. Hope it makes sense!

**Comment by shekyan** [comment]
Addressed comments.

**Comment by mikewest** [comment]
Thanks! LGTM.

**Pull request merged**